### PR TITLE
Add shopping cart

### PR DESCRIPTION
**Pull request by OleksandrSaven**

### CODE REVIEW

File: src/main/java/com/app/onlinebookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Shopping cart management", description = "Endpoints for managing shopping cart")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/api/cart")
```
Comment (holsoni) [inline src/main/java/com/app/onlinebookstore/controller/ShoppingCartController.java]: You can add "/api" path to application properties and don't write it in every controller mapping : `server.servlet.context-path=/api
`

File: src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    @Transactional
+    public CartItemDto save(CreateCartItemRequestDto requestDto) {
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(bookRepository.findById(requestDto.getBookId()).get());
+        cartItem.setQuantity(requestDto.getQuantity());
+        Long userId = userService.getAuthenticatedUser().getId();
+        Long shoppingCartId = shoppingCartRepository.findByUserId(userId).orElseThrow().getId();
```
Comment (holsoni) [inline src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java]: You aren't throwing an error here

File: src/main/java/com/app/onlinebookstore/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -46,7 +52,17 @@ public UserRegistrationResponseDto register(
...
         user.setRoles(roles);
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
         User saveUser = userRepository.save(user);
         return userMapper.toUserResponse(saveUser);
     }
+
+    @Override
+    public User getAuthenticatedUser() {
```
Comment (holsoni) [inline src/main/java/com/app/onlinebookstore/service/impl/UserServiceImpl.java:63]: Just comment: to not write this method you can pass Authentication in controllers methods to services, and there you can extract user from the authentication object: `   //Controller
@GetMapping
    public ShoppingCartDto getShoppingCart(Authentication authentication) {
        return shoppingCartService.getShoppingCart(authentication);
    }`
`//Service
 @Override
    public ShoppingCartDto getShoppingCart(Authentication authentication) {
        User currentUser = (User) authentication.getPrincipal();
        ......
`

**Review by holsoni — CHANGES_REQUESTED** [review]
Small corrections and one comment.

**Review by holsoni — APPROVED** [review]

File: src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        cartItem.setQuantity(requestDto.getQuantity());
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        Optional<User> user = userRepository.findByEmail(authentication.getName());
+        Long userId = user.get().getId();
+        setShoppingCartAndCartItems(shoppingCartRepository
+                .findByUserId(userId).get().getId(), cartItem);
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java:47]:
```suggestion
    @Override
    @Transactional
```
This method also should be transactional, as you fetching entity, setting a new field and saving it

File: src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final UserRepository userRepository;
+
+    @Override
+    @Transactional
+    public CartItemDto save(CreateCartItemRequestDto requestDto) {
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(bookRepository.findById(requestDto.getBookId()).get());
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java]: What if book does not exist?

File: src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        cartItem.setQuantity(requestDto.getQuantity());
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        Optional<User> user = userRepository.findByEmail(authentication.getName());
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java]: You can fetch user from the Authentication without calling a repository. Additionally, this is not a responsibiliti of cart item service, extract this logic to the user or authentication service

File: src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public List<CartItemDto> findCartItemsByShoppingCarts(Long id) {
+        return cartItemRepository.findCartItemsByShoppingCartId(id).stream()
+                .map(cartItemMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void delete(Long id) {
+        cartItemRepository.deleteById(id);
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/CartItemServiceImpl.java]: What if user wants to delete item which doesn't belong to him? Make sure you are securing your APIs not only by endpoint access control, but also by resource access. Make sure to add this type of security to all serivces where it is needed

File: src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    public CartItemDto save(CreateCartItemRequestDto requestDto) {
+        return cartItemService.save(requestDto);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto getShoppingCart(Authentication authentication, Pageable pageable) {
+        User currentUser = (User) authentication.getPrincipal();
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findByUserId(currentUser.getId()).orElseThrow(
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:34]: IMO, user always should have a cart. SImply create a cart if it still doesn't exist

File: src/main/resources/db/changelog/changes/10-create-cart_items-table.yaml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: shopping_cart_id
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/10-create-cart_items-table.yaml:17]: add FK

File: src/main/resources/db/changelog/changes/10-create-cart_items-table.yaml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: book_id
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/10-create-cart_items-table.yaml:24]: add FK

**Review by vovasalyha — CHANGES_REQUESTED** [review]
GJ. Let's improve it a bit

File: src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java (hunk 8 above)
Comment (OleksandrSaven) [reply src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:34]: ShoppingCart created when the user registers

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**